File: LeetCode/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        s,e=head,head
        for _ in range(n):
            if e.next!=None:
                e=e.next
            else:
                s=s.next
                return s
        while e.next!=None:
            s=s.next
            e=e.next
        s.next=s.next.next
        return head
# @lc code=end

File: LeetCode/19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */
#include <bits/stdc++.h>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *prev, *nNode, *temp;
        temp = head;
        nNode = head;
        prev = head;
        for (int i = 0; i < n; i++)
            temp = temp->next;
        while (temp != nullptr)
        {
            prev = nNode;
            temp = temp->next;
            nNode = nNode->next;
        }
        if (prev->next == nullptr)
            return nullptr;
        else if (prev == nNode)
            head = head->next;
        else
            prev->next = nNode->next;
        return head;
    }
};
// @lc code=end
